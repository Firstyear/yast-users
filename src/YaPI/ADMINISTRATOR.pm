package YaPI::ADMINISTRATOR;

use strict;
use YaST::YCP qw(:LOGGING);
use YaPI;

textdomain ("users");

# ------------------- imported modules
YaST::YCP::Import ("MailAliases");
YaST::YCP::Import ("Users");
# -------------------------------------

our $VERSION		= '1.0.0';
our @CAPABILITIES 	= ('SLES11');
our %TYPEINFO;

=item *
C<$hash Read ();>

Returns the information about system administrator (root).
Currently return hash contains the list of mail aliases.

=cut

BEGIN{$TYPEINFO{Read} = ["function",
    [ "map", "string", "any" ]];
}
sub Read {

    my $self	= shift;

    my $root_mail	= MailAliases->GetRootAlias ();
    return {} if !defined $root_mail;

    my @root_aliases	= ();
    foreach my $alias (split (/,/, $root_mail)) {
	$alias	=~ s/[ \t]//g;
	push @root_aliases, $alias;
    }

    return {
	"aliases"	=> \@root_aliases
    }
}

=item *
C<$string Write ($argument_hash);>

write the system adminstrator data. Supported keys of the argument hash are:

    "aliases"	=> list of mail aliases
    "password"	=> new password
    
Returns error message on error.

=cut

BEGIN{$TYPEINFO{Write} = ["function",
    "string",
    [ "map", "string", "any" ]];
}
sub Write {

    my $self	= shift;
    my $args	= shift;
    my $ret	= "";

    if ($args->{"aliases"} && ref ($args->{"aliases"}) eq "ARRAY") {

	my $root_mail	= join (", ", @{$args->{"aliases"}});
	if (!MailAliases->SetRootAlias ($root_mail)) {
	    # error popup
	    $ret = __("An error occurred while setting forwarding for root's mail.");
	    return $ret;
	}
    }
    if ($args->{"password"}) {
	Users->SetRootPassword ($args->{"password"});
	Users->WriteRootPassword ();
    }
    return $ret;
}

