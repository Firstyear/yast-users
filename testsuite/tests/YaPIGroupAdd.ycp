/**
 * $Id$
 */

{
    // testedfiles: Users.pm UsersCache.pm USERS.pm

    include "testsuite.ycp";

    import "YaPI::USERS";
    import "Mode";

    map R = $[
	"passwd": $[
	    "local": $[
		"users": $[
		    500: $[
			"uid":	"hh"
		    ],
		    "by_name": $[
			"hh":	500,
		    ],
		    "last_uid": 500,
		    "homes":	$[
			"/home/hh":	1,
		    ],
		    "usernames": $[
			"hh":		1,
		    ],
		    "uids": $[
			500:		1,
		    ]
		],
		"shadow": $[
		],
		"groups": $[
		    100: $[
			"cn"	:	"users",
			"gidnumber":	100,
			"type"	:	"local"
		    ],
		    "by_name": $[
			"users":	100,
		    ],
		    "gids": $[
			100:		1,
		    ],
		    "groupnames": $[
			"users":	1,
		    ]
		],
	    ],
	    "system": $[
		"users": $[
		    0:	$[
			"uid"		: "root",
			"uidnumber"	: 0
		    ],
		    "by_name": $[
			"root"		: 0
		    ],
		    "last_uid": 0,
		    "homes":	$[
		    ],
		    "usernames": $[
			"root"		: 1,
		    ],
		    "uids": $[
			0		: 1
		    ],
		],
		"shadow": $[
		],
		"groups": $[
		    0: $[
			"cn":	"root",
			"gidnumber":	0,
			"type"	:	"system"
		    ],
		    "by_name": $[
			"root"	: 0
		    ],
		    "gids": $[
			0	: 1
		    ],
		    "groupnames": $[
			"root"	: 1
		    ]
		],
	    ],
	],
	"etc" : $[
	    "fstab": [],
	    "cryptotab": [],
	    "default": $[
		"useradd": $[
		    "home":	"/tmp/local/home",
		    "groups":	"audio,video",
		    "group":	100
		]
	    ],
	],
	"target": $[
	    "stat"	: $[],
	    "size"	: 0,
	    "tmpdir"	: "/tmp/YaST"
	],
	"product": $[
		"features": $[
		    "USE_DESKTOP_SCHEDULER"	: "no",
		    "ENABLE_AUTOLOGIN"		: "false",
		    "IO_SCHEDULER"		: "",
		    "EVMS_CONFIG"		: "no"
		]
	],
    ];
    map W = $[];
    map E = $[
	"passwd" : $[
	    "init"	: true
	],
	"target": $[
	    "mkdir"	: true,
	]
    ];

    DUMP ("==========================================================");

    Mode::test = true;

    map<string,any> config_map  = $[];
    map<string,any> data_map    = $[
	"cn"	: "gg",
    ];

    DUMP ("============ add new group 'gg': ==========================");
    TEST (``(YaPI::USERS::GroupAdd (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ add new group 'gg' - done ====================");


    data_map ["userlist"]	= [ "hh1", "hh2" ];

    DUMP ("============ add new group 'gg' with first userlist as list ==");
    TEST (``(YaPI::USERS::GroupAdd (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ add new group 'gg' - done ====================");

    data_map ["userlist"]	= $[
	"hh1"	: 1,
	"hh2"	: 1
    ];

    DUMP ("============ add new group 'gg' with first userlist as map ==");
    TEST (``(YaPI::USERS::GroupAdd (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ add new group 'gg' - done ====================");


    DUMP ("============ add new group 'root' (groupname conflict): ======");

    data_map	= $[
	"cn"	: "root"
    ];

    string error = (string)
	TEST (``(YaPI::USERS::GroupAdd (config_map, data_map)), [R,W,E], 0);
    DUMP (sformat ("------------ GroupAdd return value:\n%1", error));

    DUMP ("============ add new group 'root' - done ====================");
    DUMP ("==========================================================");
}
