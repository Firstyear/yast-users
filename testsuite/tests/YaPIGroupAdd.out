Dump	==========================================================
Dump	============ add new group 'gg': ==========================
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $["by_uidnumber":$[500:$["hh":1]], "hh":$["uid":"hh"], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_uidnumber $[500:$["hh":1]]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $["by_gidnumber":$[100:$["users":1]], "gids":$[100:1], "groupnames":$["users":1], "users":$["cn":"users", "gidnumber":100, "type":"local"]]
Read	.passwd.local.groups.by_gidnumber $[100:$["users":1]]
Read	.passwd.system.users $["by_uidnumber":$[0:$["root":1]], "homes":$[], "last_uid":0, "root":$["uid":"root", "uidnumber":0], "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_uidnumber $[0:$["root":1]]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $["by_gidnumber":$[0:$["root":1]], "gids":$[0:1], "groupnames":$["root":1], "root":$["cn":"root", "gidnumber":0, "type":"system"]]
Read	.passwd.system.groups.by_gidnumber $[0:$["root":1]]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash_output "echo 'gg' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":"gg"]
Execute	.target.bash "/bin/cp /etc/group /etc/group.YaST2save" 0
Write	.passwd.groups $["local":$["by_gidnumber":$["100":$["users":"1"]], "gg":$["cn":"gg", "gidnumber":"1000", "modified":"added", "org_cn":"gg", "org_gidnumber":"1000", "plugins":[], "type":"local", "userlist":$[], "what":"add_group"], "gids":$["100":"1"], "groupnames":$["users":"1"], "users":$["cn":"users", "gidnumber":"100", "type":"local"]], "system":$["by_gidnumber":$["0":$["root":"1"]], "gids":$["0":"1"], "groupnames":$["root":"1"], "root":$["cn":"root", "gidnumber":"0", "type":"system"]]] true
Execute	.target.bash "/usr/sbin/nscd -i group" 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $["custom_groups":["local"], "custom_users":["local"], "dont_warn_when_nisserver_notdes":false, "dont_warn_when_uppercase":false] true
Return	
Dump	============ add new group 'gg' - done ====================
Dump	============ add new group 'gg' with first userlist as list ==
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $["by_uidnumber":$[500:$["hh":1]], "hh":$["uid":"hh"], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_uidnumber $[500:$["hh":1]]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $["by_gidnumber":$[100:$["users":1]], "gids":$[100:1], "groupnames":$["users":1], "users":$["cn":"users", "gidnumber":100, "type":"local"]]
Read	.passwd.local.groups.by_gidnumber $[100:$["users":1]]
Read	.passwd.system.users $["by_uidnumber":$[0:$["root":1]], "homes":$[], "last_uid":0, "root":$["uid":"root", "uidnumber":0], "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_uidnumber $[0:$["root":1]]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $["by_gidnumber":$[0:$["root":1]], "gids":$[0:1], "groupnames":$["root":1], "root":$["cn":"root", "gidnumber":0, "type":"system"]]
Read	.passwd.system.groups.by_gidnumber $[0:$["root":1]]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash_output "echo 'gg' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":"gg"]
Execute	.target.bash "/bin/cp /etc/group /etc/group.YaST2save" 0
Write	.passwd.groups $["local":$["by_gidnumber":$["100":$["users":"1"]], "gg":$["cn":"gg", "gidnumber":"1000", "modified":"added", "org_cn":"gg", "org_gidnumber":"1000", "plugins":[], "type":"local", "userlist":$["hh1":"1", "hh2":"1"], "what":"add_group"], "gids":$["100":"1"], "groupnames":$["users":"1"], "users":$["cn":"users", "gidnumber":"100", "type":"local"]], "system":$["by_gidnumber":$["0":$["root":"1"]], "gids":$["0":"1"], "groupnames":$["root":"1"], "root":$["cn":"root", "gidnumber":"0", "type":"system"]]] true
Execute	.target.bash "/usr/sbin/nscd -i group" 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $["custom_groups":["local"], "custom_users":["local"], "dont_warn_when_nisserver_notdes":false, "dont_warn_when_uppercase":false] true
Return	
Dump	============ add new group 'gg' - done ====================
Dump	============ add new group 'gg' with first userlist as map ==
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $["by_uidnumber":$[500:$["hh":1]], "hh":$["uid":"hh"], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_uidnumber $[500:$["hh":1]]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $["by_gidnumber":$[100:$["users":1]], "gids":$[100:1], "groupnames":$["users":1], "users":$["cn":"users", "gidnumber":100, "type":"local"]]
Read	.passwd.local.groups.by_gidnumber $[100:$["users":1]]
Read	.passwd.system.users $["by_uidnumber":$[0:$["root":1]], "homes":$[], "last_uid":0, "root":$["uid":"root", "uidnumber":0], "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_uidnumber $[0:$["root":1]]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $["by_gidnumber":$[0:$["root":1]], "gids":$[0:1], "groupnames":$["root":1], "root":$["cn":"root", "gidnumber":0, "type":"system"]]
Read	.passwd.system.groups.by_gidnumber $[0:$["root":1]]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash_output "echo 'gg' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":"gg"]
Execute	.target.bash "/bin/cp /etc/group /etc/group.YaST2save" 0
Write	.passwd.groups $["local":$["by_gidnumber":$["100":$["users":"1"]], "gg":$["cn":"gg", "gidnumber":"1000", "modified":"added", "org_cn":"gg", "org_gidnumber":"1000", "plugins":[], "type":"local", "userlist":$["hh1":"1", "hh2":"1"], "what":"add_group"], "gids":$["100":"1"], "groupnames":$["users":"1"], "users":$["cn":"users", "gidnumber":"100", "type":"local"]], "system":$["by_gidnumber":$["0":$["root":"1"]], "gids":$["0":"1"], "groupnames":$["root":"1"], "root":$["cn":"root", "gidnumber":"0", "type":"system"]]] true
Execute	.target.bash "/usr/sbin/nscd -i group" 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $["custom_groups":["local"], "custom_users":["local"], "dont_warn_when_nisserver_notdes":false, "dont_warn_when_uppercase":false] true
Return	
Dump	============ add new group 'gg' - done ====================
Dump	============ add new group 'root' (groupname conflict): ======
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $["by_uidnumber":$[500:$["hh":1]], "hh":$["uid":"hh"], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_uidnumber $[500:$["hh":1]]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $["by_gidnumber":$[100:$["users":1]], "gids":$[100:1], "groupnames":$["users":1], "users":$["cn":"users", "gidnumber":100, "type":"local"]]
Read	.passwd.local.groups.by_gidnumber $[100:$["users":1]]
Read	.passwd.system.users $["by_uidnumber":$[0:$["root":1]], "homes":$[], "last_uid":0, "root":$["uid":"root", "uidnumber":0], "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_uidnumber $[0:$["root":1]]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $["by_gidnumber":$[0:$["root":1]], "gids":$[0:1], "groupnames":$["root":1], "root":$["cn":"root", "gidnumber":0, "type":"system"]]
Read	.passwd.system.groups.by_gidnumber $[0:$["root":1]]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash_output "echo 'root' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":"root"]
Return	There is a conflict between the entered
Dump	------------ GroupAdd return value:
Dump	There is a conflict between the entered
Dump	group name and an existing group name.
Dump	Try another one.
Dump	============ add new group 'root' - done ====================
Dump	==========================================================
